sdk/core/System.ClientModel/src/Message/PipelineRequestHeaders.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace System.ClientModel.Primitives;
+
+internal class PipelineRequestHeaders : PipelineMessageHeaders
+{
+    private ArrayBackedPropertyBag<IgnoreCaseString, object> _headers;
+
+    public override bool Remove(string name)
+        => _headers.TryRemove(new IgnoreCaseString(name));
+
+    public override void Set(string name, string value)
+        => _headers.Set(new IgnoreCaseString(name), value);
+
+    public override void Add(string name, string value)
+    {
+        if (_headers.TryAdd(new IgnoreCaseString(name), value, out object? currentValue))
+        {
+            return;
+        }
+
+        switch (currentValue)
+        {
+            case string stringValue:
+                _headers.Set(new IgnoreCaseString(name), new List<string> { stringValue, value });
```

Comment:
Pri 2: I wonder if we could/should not collect multiple values in a single string (comma separated). It seems like it's permitted by the HTTP spec, though I am not sure how HttpClient would deal with that, i.e. if we add a header to HttpClientMessage where the header value is actually multiple values comma separated. See the following in HTTP spec (https://www.rfc-editor.org/rfc/rfc2616#page-31):

Multiple message-header fields with the same field-name MAY be
   present in a message if and only if the entire field-value for that
   header field is defined as a comma-separated list [i.e., #(values)].
   It MUST be possible to combine the multiple header fields into one
   "field-name: field-value" pair, without changing the semantics of the
   message, by appending each subsequent field-value to the first, each
   separated by a comma. The order in which header fields with the same
   field-name are received is therefore significant to the
   interpretation of the combined field value, and thus a proxy MUST NOT
   change the order of these field values when a message is forwarded.
   
   If we did that, we could remove the APIs (below) that let users read IEnumeralbe<T> values, and we would not have to allocate collection/lists here. This would make the implementation so much simpler.

Reply:
Interesting!  This would take a little work to investigate both from perf and back-compat perspectives, so filing https://github.com/Azure/azure-sdk-for-net/issues/41228 to track.